### ISSUE TITLE
Clean up compliance tests and migrate to more unit tests

### ISSUE THREAD

**Issue by jeenbroekstra** [eclipse/rdf4j#1236]
Several test implementations in the various compliance modules are overlapping in coverage, and/or really should be refactored to proper unit tests (in src/test/java of the module/class being tested, instead of a separate compliance module).

A difficulty is that any test that needs a sail back-end store can introduce a circular dependency in maven when added to it main module instead of compliance. We should look into a mocking/stub setup to help alleviate this. 

Followup of #1230

**Comment by jeenbroekstra** [comment]
Relevant: https://maven.apache.org/plugins/maven-jar-plugin/examples/create-test-jar.html

**Comment by jeenbroekstra** [comment]
For the Model API, I will get rid of the compliance module and the testsuite completely, and just turn the whole thing into unit tests. While we're at it we'll replace the poorly integrated code using Apache commons collections tests with Google Guava test code for Java collections. 

CQ for the guava testlib logged here : https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19292

**Comment by jeenbroekstra** [comment]
CQ approved

**Comment by hmottestad** [comment]
I've had some issues with debugging the SHACL compliance tests in IntelliJ. Will this change make it possible to debug a single test?

**Comment by jeenbroekstra** [comment]
@hmottestad I'm not immediately sure. What you describe is often caused by using tests that are encoded in resource files - in such cases the unit test runner only knows _at runtime_ which tests exist, so it can't "zoom in" on a single test easily. I'm not proposing to refactor that complete setup as part of this cleanup as well (though that doesn't mean I'm against doing it, just that I want to keep this issue manageable in size ;)).

**Comment by jeenbroekstra** [comment]
The proposed new setup is as follows:

1. all compliance tests will be moved into the rdf4j-testsuite repo. This removes circular dependencies between the other repos and the testsuites, and in addition makes quicker test feedback loops possible on the compliance tests. 
2. various tests that were in the past designed as compliance are, in fact, unit tests, and will be moved _from_ the testsuites/compliance modules to their respective functional modules. For example, the model tests, and the queryresultio tests.

A downside of this approach is that compliance/integration testing will not be run as part of the verification cycle for rdf4j, rdf4j-storage, and rdf4j-tools. It will instead be part of the rdf4j-testsuite compile/verify cycle. In order to mitigate this it's important that we migrate as many tests as possible to the level of unit tests (_without_ reintroducing cyclic dependencies, however).

**Comment by hmottestad** [comment]
Are you proposing that we don’t even call the compliance tests in the respective repos? Today, in RDF4J-storage, there is a compliance module which extends a bunch of tests. 

I’m very worried if all the tests aren’t run automatically on PRs in Jenkins! 

I propose the following:
 - all tests that are not shared by more than one repo are moved into that repo (compliance or not)
 - all modules that cause cyclic dependencies are moved into single repos, and possibly merged into single modules

I must admit, that after we split up RDF4J into multiple repos, I’ve become a big fan of single repos. For build performance we could instead try out gradle. 

Also, I find it very annoying that when a compliance test fails, I need to go to the testsuite github and search up the file in resources that was loaded as part of the test.

**Comment by jeenbroekstra** [comment]
A possible change is to move the test suites and the compliance tests into the corresponding repos, and get rid of rdf4j-testsuites altogether.  I didn't immediately want to go that radical but perhaps we should.

**Comment by jeenbroekstra** [comment]
What I hadn't thought of when I proposed that is that we _can't_ do that. The testsuites and compliance tests all have multiple dependencies on modules in both rdf4j and rdf4j-storage (and even rdf4j-tools).

For example, the rio-testsuite tests use the memory sail. This means we can't compile and execute them until _after_ the memory sail module has been built. Therefore we can not put these tests in the rdf4j repo, because that repo gets built _before_ the rdf4j-storage repo (which contains the memory sail). This is precisely why we currently have this convoluted setup with only including the compliance modules in the maven build when executing tests, and having jenkins first do a deployment _without_ tests and only _afterwards_ running the compliance tests. 

As a first pass I will continue with my earlier proposal: move all compliance tests to the testsuite, and migrate as many non-complex tests as possible back to proper unit tests included in each functional module itself. 

But I'm considering further changes. One possible change is to make sure that dependencies on other modules in the testsuite are not against the latest snapshot, but against a fixed release - the thinking being that it shouldn't matter for, for example, the rio testsuite which particular version of the memory sail it uses - it just needs a store to keep some data in. If we do that, we can have these tests moved back into the rdf4j repo, because they no longer rely on the latest rdf4j-storage repo being built. 

I also want to have a think about the possibility of moving back to a single repo. There's downsides to that, but a big upside is that a single maven process has access to the latest snapshot builds of every part of the framework, without having to rely on snapshots first being deployed to sonatype.

**Comment by jeenbroekstra** [comment]
PRs open for a complete overhaul. In summary, what we changed:

- all testsuite modules from rdf4j-testsuite have been moved to either rdf4j-storage or rdf4j repos. 
- several compliance tests have been changed to become (abstract) unit tests. E.g. `RepositoryConnectionTest` is now a unit test packaged in the `repository-api` module, and is re-used by unit tests in, for example, the `repository-sail` module. 
- compliance tests that rely on "downstream" modules now rely on a fixed released version (e.g. the sparql repository compliance tests rely on spinning up a rdf4j server version 2.5.0).
- replaced use logback in testing with slf4j-simple, which is a more lightweight stdout logger (I may have missed a few though)

Further cleanup may be necessary but this is the gist of it. Note that due to these changes we will need to adjust our Jenkins setup significantly.

A major benefit of this approach is that we no longer have any circular dependencies, plus that all tests effectively live in the same repo as the code they're testing.
